src/lib/pickles_types/hlist.mli

```diff
@@ -0,0 +1,920 @@
+module Tuple5 : functor (F : T3) (G : T3) (H : T3) (I : T3) (J : T3) -> sig
+  type ('a, 'b, 'c) t =
+    ('a, 'b, 'c) F.t
+    * ('a, 'b, 'c) G.t
+    * ('a, 'b, 'c) H.t
+    * ('a, 'b, 'c) I.t
+    * ('a, 'b, 'c) J.t
+end
+
+(** {2 Argument selectors} *)
+
+(** The following Functions just return one of their arguments. *)
+
+(** Always return first argument: [let fst a _ = a] *)
+module Arg1 : sig
+  type ('a, _, _) t = 'a
+end
+
+(** Always return second argument: [let snd _ b = b] *)
+module Arg2 : sig
+  type (_, 'a, _) t = 'a
+end
+
+(** {2 Application transformers} *)
+
+(** The following higher-order functions composes function applications in
+    specific ways.
+*)
+
+(** Compose one binary function [F] and two unary functions [X] and [Y] into
+    a binary function that applies [X] (resp. [Y]) to its first (resp. second)
+    argument before passing both to [F]:
+    [let apply2 f x y = fun a b -> f (x a) (y b)]
+*)
+module Apply2 : functor (F : T2) (X : T1) (Y : T1) -> sig
+  type ('a, 'b) t = ('a X.t, 'b Y.t) F.t
+end
+
+(** Transforms a binary function into an unary one by duplicating its argument.
+    [let dup f = fun x -> f x x]
+*)
+module Dup : functor (F : T2) -> sig
+  type 'a t = ('a, 'a) F.t
+end
+
+(** Reification of list length with Peano integers, see {!Hlist0.Length}. *)
+module Length = Hlist0.Length
+
+(** {1:lists Heterogeneous lists} *)
+
+(** {2 Over one type parameter} *)
+
+(**
+   The following modules are named as with the following scheme:
+   - Module H[X] is the type of (and operations on) heterogeneous lists whose
+     content type varies over [X] different type parameters. See {!H1} for a
+     detailed example.
+   - Module H[X]_[Y] is the type of (and operations on) heterogeneous lists whose
+     content type varies over [X] different type parameters from cell to cell,
+     but also varies from list to list over [Y] other homogeneous type
+     parameters. See {!Hlist0.H1_1} for a detailed example.
+*)
+
+(** Operations on heterogeneous lists whose content type varies over a single
+    type parameter. Read {!Hlist0.H1} first.
+
+    Most operations are functors that take a module [F] as a first argument.
+    This module is a type function which maps the varying type parameter to the
+    actual contained type.
+
+    In all examples below, we will assume that [F] has the following definition:
+    {[module F = struct
+        type 'a t = 'a option
+      end
+    ]}
+*)
+module H1 : sig
+  (** Core datatype, see {!Hlist0.H1}. *)
+  module T = Hlist0.H1
+
+  (** Iterate over a heterogeneous list with a polymorphic function over the
+      varying type parameter.
+
+      The second functor parameter is a module containing that polymorphic
+      function. For example, with
+      {[module C = struct
+          let f (x : 'a option) = assert (Option.is_some x)
+        end
+      ]}
+      [H1.Iter.(F)(C).f l] will iterate over the list [l] and check that every
```

Comment:
`check that every cell is some value` seems weird here? Surely it's `apply C.f to each value in order`?

Reply:
Yeah, I was trying to say `check that every value is Some x`, I'll rephrase.